### PR TITLE
Allow None buildtype

### PR DESCRIPTION
**Pull request by PureTryOut**
Some distributions, in my case Alpine Linux, require building CMake packages with the
"None" buildtype as this makes packages use compiler flags required by
the distribution.

See a discussion with the Arch Linux maintainer:
https://github.com/archlinux/svntogit-community/commit/da37cf256156dead98e5a62219ab4006167986f9#r53047679

And with one of the Alpine Linux packagers:
https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/23363#note_168960

### CODE REVIEW

**Comment by Be-ing** [comment]
Please rebase on the 2.3 branch.

**Comment by PureTryOut** [comment]
Done

**Comment by ronso0** [comment]
I think this is your first contribution (someone please double-check) we need you to sign the [contributor agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ) and thereby allows Mixxx to include your changes.

**Comment by PureTryOut** [comment]
It is indeed and I just signed it :+1:

**Comment by Holzhaus** [comment]
Could you explain what builds flags in particular you need? Mixxx is very picky about compiler flags.  For example, we already the issue that Debian changed the build flags on ARM and disabled NEON to match their distribution policy, which made Mixxx completely unusable on all ARM machines (IIRC). 

I think we disallow other build types by design for that reason. I'm open to remove the check, but it should probably print a warning that using custom flags is discouraged.

**Comment by PureTryOut** [comment]
> Could you explain what builds flags in particular you need?

If you're asking that to maybe make the Release build type or whatever use those flags, that's not useful to us. We can't rely on very upstream setting the flags we require correctly, it _has_ to be enforced on the distribution level.

I'd be fine with adding a warning, it's up to the distribution packagers anyway to make sure the package works as expected.

**Comment by PureTryOut** [comment]
I've added a warning, please let me know if this is the right approach and/or message.

**Comment by daschuer** [comment]
> Could you explain what builds flags in particular you need?

Could you answered this question explicit?

I think we are all in the same page and want to deliver the best user experience with Mixxx in Linux. I this project we went to some cycles of improving the build flags and we are now convinced that the current setting are optimal.

> We can't rely on very upstream setting the flags we require correctly, it has to be enforced on the distribution level.

I think In the Mixxx case you can rely. But let's find out if this is the case. 

I think the main issue is when it comes to user support. We can't really support Mixxx builds that are build with Distribution default flags that we don't know.

**Comment by Holzhaus** [comment]
> I think the main issue is when it comes to user support. We can't really support Mixxx builds that are build with Distribution default flags that we don't know.

It's the classic problem. The same goes for distro-specific patches. I guess we could merge this PR and just tell all users that use a mixxx version not packaged by ourselves to go to their distro's bug tracker instead.

**Comment by daschuer** [comment]
Yes, but that the user experience I like to avoid. 

My hope is that we can use the discussion here to have a solution without that for Alpine Linux at least.

**Comment by PureTryOut** [comment]
> tell all users that use a mixxx version not packaged by ourselves to go to their distro's bug tracker instead.

That's how it should go anyway. If you use a package from distribution repositories, you have to report bugs to the distribution issue trackers. If then the maintainers/packagers can reproduce it but not immediately fix it themselves, then they will have to report it upstream. I expect nothing else personally and nobody should really.

> Could you explain what builds flags in particular you need?

As said, no we can not rely on individual packages. We're not in a habit to make exceptions either; you might set the stuff we require now, but who says you won't change them in the future? I understand why you would want everybody to use specific options, but it's not your responsibility to fix issues that come from changing those, let us distro packagers do that :wink: 

Anyway, we set them [here](https://gitlab.alpinelinux.org/alpine/abuild/-/blob/master/abuild.conf).

**Comment by daschuer** [comment]
That's a pitty. I sound like you treat following dist policy as more important than delivering a build useful for the users the users? 
What I can spot, you will use only
-Os -fomit-frame-pointer
Really? Optimize for size? 

We use:
-O3 -ffast-math -funroll-loops -fomit-frame-pointer -mtune=generic

This Makes the EQs and some effects usable without hitting 100 % CPU load when phased out. 

If you optimize other CPU hungry libraries like soundtouch and rubberband also only for size the result becomes even worse.

**Comment by PureTryOut** [comment]
> It sound like you treat following dist policy as more important than delivering a build useful for the users the users?

Well for one, it's not my decision; it's distribution policy. Secondly, following our own compilation options does not mean the build won't be useful to our users. Mixxx very much works fine on Alpine Linux and always has, we only started caring about this now because your build system changed to CMake (which is a good thing!) which forces different defaults using that build type. Again, it's our own responsibility to make sure the package works with the compilation options we provide, not yours.

> Really? Optimize for size?

Yes, Alpine Linux is not a traditional desktop distribution. It's used a lot in embedded scenarios and in Docker containers, which definitely require optimizing for size. That said, it's a "general purpose" distribution so you can use it as a desktop distro just fine, and that's what at least I myself do.

**Comment by daschuer** [comment]
By the way, you build the 2.2.4 version with these flags:
 
'-Os', '-pipe', '-Wall', '-Wextra', '-g', '-fPIC', '-O3', '-ffast-math', '-funroll-loops', '-fomit-frame-pointer', '-mtune=generic', '-msse2', '-mfpmath=sse', '-pthread'

No need for a change.

**Comment by Be-ing** [comment]
Our custom compiler options are set based on the OPTIMIZE option, not the CMake build type:
```
elseif(GNU_GCC OR LLVM_CLANG)
  if(NOT OPTIMIZE STREQUAL "off")
    # Common flags to all optimizations.
    # -ffast-math will prevent a performance penalty by denormals
    # (floating point values almost Zero are treated as Zero)
    # unfortunately that work only on 64 bit CPUs or with sse2 enabled
    # The following optimisation flags makes the engine code ~3 times
    # faster, measured on a Atom CPU.
    add_compile_options(
      -O3
      -ffast-math
      -funroll-loops
    )
    # set -fomit-frame-pointer when we don't profile and are not using
    # Clang sanitizers.
    # Note: It is only included in -O on machines where it does not
    # interfere with debugging
    if(NOT PROFILING AND NOT CLANG_SANITIZERS)
      add_compile_options(-fomit-frame-pointer)
    endif()
```
So I think the None build type should be okay?

**Comment by Be-ing** [comment]
I tried building with build type None and this is the command line:
```
[145/809] ccache /usr/bin/c++ -DHAVE_INET_ATON -DHAVE_INET_PTON -DHAVE_UNISTD_H -DMIXXX_BUILD_RELEASE -DNDEBUG -DQT_CONCURRENT_LIB -DQT_CORE_LIB -DQT_DBUS_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -DQT_OPENGL_LIB -DQT_SCRIPTTOOLS_LIB -DQT_SCRIPT_LIB -DQT_SQL_LIB -DQT_SVG_LIB -DQT_TABLET_SUPPORT -DQT_TESTCASE_BUILDDIR=\"/home/be/sw/mixxx/build\" -DQT_TESTLIB_LIB -DQT_USE_QSTRINGBUILDER -DQT_WIDGETS_LIB -DQT_X11EXTRAS_LIB -DQT_XML_LIB -D__BATTERY__ -D__BROADCAST__ -D__BULK__ -D__FAAD__ -D__FFMPEG__ -D__HID__ -D__KEYFINDER__ -D__LILV__ -D__LINUX__ -D__MAD__ -D__MODPLUG__ -D__MP4V2__ -D__OPUS__ -D__QTKEYCHAIN__ -D__SNDFILE__ -D__SQLITE3__ -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -D__UNIX__ -D__VINYLCONTROL__ -D__WV__ -Dx86_64 -Imixxx-lib_autogen/include -I../src -Isrc -I/usr/include/ffmpeg -isystem ../lib/fidlib -isystem ../lib/googletest/googletest/include -isystem /home/be/local/include -isystem ../lib/portaudio -isystem ../lib/rigtorp/SPSCQueue/include -isystem ../lib/qtscript-bytearray -isystem ../lib/replaygain -isystem ../lib/reverb -isystem /usr/include/glib-2.0 -isystem /usr/lib64/glib-2.0/include -isystem ../lib/libshout-idjc/include -isystem /usr/include/opus -isystem ../lib/kaitai -isystem ../lib/mp3guessenc-0.27.4 -isystem /usr/include/qt5 -isystem /usr/include/qt5/QtConcurrent -isystem /usr/include/qt5/QtCore -isystem /usr/lib64/qt5/mkspecs/linux-g++ -isystem /usr/include/qt5/QtGui -isystem /usr/include/qt5/QtNetwork -isystem /usr/include/qt5/QtOpenGL -isystem /usr/include/qt5/QtWidgets -isystem /usr/include/qt5/QtScript -isystem /usr/include/qt5/QtScriptTools -isystem /usr/include/qt5/QtSql -isystem /usr/include/qt5/QtSvg -isystem /usr/include/qt5/QtTest -isystem /usr/include/qt5/QtXml -isystem /usr/include/qt5/QtX11Extras -isystem /usr/include/qt5/QtDBus -isystem ../lib/qm-dsp -isystem ../lib/qm-dsp/include -isystem /usr/include/taglib -isystem /usr/include/libupower-glib -isystem /usr/include/lilv-0 -isystem ../lib/hidapi/hidapi -isystem /usr/include/libusb-1.0 -isystem ../lib/xwax -isystem /usr/include/wavpack -fdiagnostics-color=auto -pipe -O3 -ffast-math -funroll-loops -fomit-frame-pointer -mtune=generic -Wall -Wextra -Woverloaded-virtual -Wfloat-conversion -Werror=return-type -fPIC -pthread -std=gnu++17 -MD -MT CMakeFiles/mixxx-lib.dir/src/control/controlpushbutton.cpp.o -MF CMakeFiles/mixxx-lib.dir/src/control/controlpushbutton.cpp.o.d -o CMakeFiles/mixxx-lib.dir/src/control/controlpushbutton.cpp.o -c ../src/control/controlpushbutton.cpp
```

`-pipe -O3 -ffast-math -funroll-loops -fomit-frame-pointer -mtune=generic -Wall -Wextra -Woverloaded-virtual -Wfloat-conversion -Werror=return-type -fPIC -pthread -std=gnu++17 -MD -MT` so I think the None build type is okay.

**Comment by daschuer** [comment]
> So I think the None build type should be okay?

None is not considered in our optimization logic, we need to check that. But for what?

**Comment by Be-ing** [comment]
> But for what?

Because packagers expect the None build type to work. Let's not make extra work for them.

**Comment by daschuer** [comment]
Building already works without setting CMAKE_BUILD_TYPE so there is no extra work for the package maintainer. 
The issue here is that @PureTryOut Wants to stop our CMakeList.txt from setting CXX_FLAGS.

**Comment by Be-ing** [comment]
My understanding, which very well could be wrong because I have not tested this, is that the important options are `-ffast-math` and `-funroll-loops`. @daschuer I think you have experimented with compiler options before. Can you summarize what you found and/or refer to the old discussion about those? I think that would help move this conversation forward.

**Comment by PureTryOut** [comment]
> The issue here is that @PureTryOut Wants to stop our CMakeList.txt from setting CXX_FLAGS.

Not entirely true, I mainly want CMake to respect our CXXFLAGS first and foremost. Any extra options are fine, it's just that our flags should take priority. I'm not enough of a CMake guru to be able to confirm having the None build type accomplishes this, but it's distribution policy to have it like that to stop CMake from doing stuff on it's own we don't want.

That said, I'm not the maintainer of this package on Alpine Linux, I'm just a packager who happened to recently find this software and wanted to upgrade it to it's latest release. @kaniini however is the maintainer and knows more about this stuff, they can probably chip in here.

**Comment by daschuer** [comment]
In the 2.2.4 build script you set -Os but it is replaced by the later set -O3. I have no experienced how Mixxx behaves with -Os and what does it actually mean for the size. 
Our current deb file is 17 MB. The unpacked data are 46 MB which contains the mixxx binary of 14.1 MB
How big is the binary compiled with -Os?

**Comment by Be-ing** [comment]
I think we can merge this as it is for 2.3.1 and continue discussing compiler flags for packaging. @daschuer is this okay?

File: CMakeLists.txt [hunk 1]
```diff
@@ -53,6 +53,8 @@ if(NOT CMAKE_CONFIGURATION_TYPES)
     set(CMAKE_BUILD_TYPE "${default_build_type}" CACHE STRING "Choose the type of build." FORCE)
     # Set the possible values of build type for cmake-gui
     set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "RelWithDebInfo")
+  elseif(CMAKE_BUILD_TYPE MATCHES "^(None)$")
+    message(WARNING "Building with custom compiler flags is discouraged and can cause unexpected problems.")
```
Comment (daschuer) [inline CMakeLists.txt:57]: Setting CMAKE_BUILD_TYPE=None does not allow to Build with custom compile flags.

**Review by daschuer — CHANGES_REQUESTED** [review]
There is no reason to clutter our CMakeLists.txt with a custom build type "None" that does not fix the original issue and is not considered in the rest of the file. 
I think this can be closed.

**Comment by Holzhaus** [comment]
> There is no reason to clutter our CMakeLists.txt with a custom build type "None" that does not fix the original issue and is not considered in the rest of the file. 
> I think this can be closed. 

I agree that explicitly adding a `None` build time would be weird. I suppose almost no project does this, because it's not a standard build type name.

Instead, I suggest we should check if the build type is either `Release`, `Debug`, `RelWithDebInfo` and `MinSizeRel`. If not, let's print a warning but allow the user to continue.

We could also consider to display a note in the about dialog in such cases, e. g. "This build is unsupported by the upstream Mixxx project. Please report bugs in your distribution's bug tracker, not the Mixxx bugtracker."

**Comment by Be-ing** [comment]
> We could also consider to display a note in the about dialog in such cases, e. g. "This build is unsupported by the upstream Mixxx project. Please report bugs in your distribution's bug tracker, not the Mixxx bugtracker."

I don't like this idea. We will miss out on bugs in Mixxx if we do this.

**Comment by daschuer** [comment]
> Instead, I suggest we should check if the build type is either Release, Debug, RelWithDebInfo and MinSizeRel. If not, let's print a warning but allow the user to continue.

What could be the reason for us to lower the barrier? 

Currently it errors fatal in that case, because it is a case, we have not considered and tested. It most likely happens because a user misspells RelWithDebInfo or friends, where a fatal error is the best responds. 

The reason why this PR exist is to build Mixxx with custom flags but a non supported None build type is not a solution anyway. 

Bypassing our build flags is also something we should make as hard as possible, because we don't want support that -> fatal error. If one insists he is free to patch the CMakeList as he can patch or any other file in Mixxx.

**Comment by kaniini** [comment]
I think in Alpine's case, mixxx 2.2.4 was built with -O3, because I didn't muck with the build system... so I can't give much comparison.
